src/Orchard/Mvc/MvcModule.cs

```diff
@@ -24,37 +27,25 @@ public class MvcModule : Module {
             moduleBuilder.Register(UrlHelperFactory).As<UrlHelper>().InstancePerDependency();
         }
 
-        private static bool IsRequestValid() {
-            if (HttpContext.Current == null)
-                return false;
-
-            try {
-                // The "Request" property throws at application startup on IIS integrated pipeline mode.
-                var req = HttpContext.Current.Request;
-            }
-            catch (Exception ex) {
-                if (ex.IsFatal()) {
-                    throw;
-                }
-
-                return false;
-            }
-
-            return true;
-        }
-
         static HttpContextBase HttpContextBaseFactory(IComponentContext context) {
-            if (IsRequestValid()) {
-                return new HttpContextWrapper(HttpContext.Current);
-            }
+
+            var httpContext = context.Resolve<IHttpContextAccessor>().Current();
+            if (httpContext != null)
+                return httpContext;
 
             var siteService = context.Resolve<ISiteService>();
+            var prefix = context.Resolve<ShellSettings>().RequestUrlPrefix;
 
             // Wrapping the code accessing the SiteSettings in a function that will be executed later (in HttpContextPlaceholder),
             // so that the RequestContext will have been established when the time comes to actually load the site settings,
             // which requires activating the Site content item, which in turn requires a UrlHelper, which in turn requires a RequestContext,
             // thus preventing a StackOverflowException.
-            var baseUrl = new Func<string>(() => siteService.GetSiteSettings().BaseUrl);
+            var baseUrl = new Func<string>(() => {
+                var url = siteService.GetSiteSettings().BaseUrl;
+                return !string.IsNullOrWhiteSpace(url) ? !string.IsNullOrWhiteSpace(prefix) ?

```

Comment:
When running on a background task, you definitely don't want to get the BaseUrl from the site settings, as this would cause an infinite loop (stopped by the Lazy type because it checks for accessing itself while creating its value). But you do need a base url when creating a fake HttpContext for HttpContext.Current for rendering views that use Html Helpers. If the Request object has no valid URL (even if it's a fake one) kittens will die a horrible death. I've seen it happen.